bookmarks/src/commonMain/kotlin/catchup/bookmarks/BookmarkRepository.kt

```diff
@@ -0,0 +1,113 @@
+package catchup.bookmarks
+
+import app.cash.sqldelight.Query
+import app.cash.sqldelight.Transacter
+import app.cash.sqldelight.coroutines.asFlow
+import app.cash.sqldelight.db.QueryResult
+import app.cash.sqldelight.db.SqlCursor
+import catchup.bookmarks.db.Bookmark
+import catchup.bookmarks.db.CatchUpDatabase
+import catchup.service.api.CatchUpItem
+import catchup.service.api.toCatchUpItem
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
+import kotlinx.datetime.Clock.System
+import kotlinx.datetime.Instant
+
+interface BookmarkRepository {
+  fun addBookmark(id: Long, timestamp: Instant = System.now())
+
+  fun removeBookmark(id: Long)
+
+  fun isBookmarked(id: Long): Flow<Boolean>
+
+  fun bookmarksCountFlow(): Flow<Long>
+
+  // Exposed to create a PagingSource
+  fun bookmarksCountQuery(): Query<Long>
+
+  fun bookmarksTransacter(): Transacter
+
+  fun bookmarksQuery(limit: Long, offset: Long): Query<CatchUpItem>
+}
+
+internal class BookmarkRepositoryImpl(private val database: CatchUpDatabase) : BookmarkRepository {
+  private val scope = CoroutineScope(Dispatchers.IO)
+
+  // Maintain an in-memory cache of all the bookmarks
+  private val bookmarks = MutableStateFlow(LinkedHashSet<Long>())
+
+  init {
+    scope.launch {
+      val idsFlow =
+        database.transactionWithResult { database.bookmarksQueries.bookmarkIds().asFlow() }
+      idsFlow.collect { query ->
+        // Preserve order
+        bookmarks.emit(query.executeAsList().mapTo(LinkedHashSet(), Bookmark::id))
+      }
+    }
+  }
+
+  override fun addBookmark(id: Long, timestamp: Instant) {
+    scope.launch { database.transaction { database.bookmarksQueries.addBookmark(id, timestamp) } }
```

Comment:
This API probably wants to be a `suspend fun`! In many scenarios the caller can and should safely own this, and a `suspend fun` still opens the door for kicking it over to a longer lived scope if the caller really does need that

If you feel strongly opinionated that this is the right API shape, then `scope.launch` *still* isn't the right choice. Use a channel instead:

```kotlin
data class AddBookmark(val id: Long, val timestamp: Instant)

private val addBookmarks = Channel<AddBookmark>(BUFFERED)

init {
  scope.launch {
    for (item in addBookmarks) {
      database.transaction { database.bookmarksQueries.addBookmark(item.id, item.timestamp) }
    }
  }
}

override fun addBookmark(id: Long, timestamp: Instant) {
  addBookmarks.sendOrFail(AddBookMark(id, timestamp))
}
```

You might say, "I don't need a queue! Why did you make me build a queue?" But you already had a queue: it was the dispatcher, and its failure mode was "my dispatcher is full of hung bookmark add operations".

Reply:
my concern about making it suspend is that I don't fully understand what that communicates to the caller. What happens if they cancel the add? Is that something the UI gets to really decide? Does the add() always work but the UI gets to know if it successfully handed it off (synchronously) to the internal transaction handler?

Reply:
* When the `suspend fun` write completes successfully, the caller knows the write succeeded
* If the write fails, the caller will get an exception (which is good)
* If the caller cancels the `addBookmark` call before it completes (e.g. by navigating away), then the write will get discarded
* If the caller wants to navigate away and not cancel the write, they can still accomplish that by doing exactly what your existing code here does

All of these are valid choices, which is why I tend to prefer suspend funs for this type of thing.